### PR TITLE
Log all different failed and recovered reconnects to backends

### LINKED ISSUE #351: Log reconnects
At the moment, Icinga DB only logs that it will retry reconnecting to the database (and Redis?). I think it makes sense to also log reconnects then. We could add a struct for optional args to the `retry.WithBackoff()` function which provides settings for `OnError` and `OnRetry` callbacks. The timeout should be moved there as well then.

### PR DESCRIPTION
**Pull request by Al2Klimov**
E.g. the first "connection refused", the first "hostname mismatch" and the all-clear.

fixes #351

### CODE REVIEW

**Comment by Al2Klimov** [comment]
```
2021-09-22T16:28:03.134+0200    WARN    config/redis.go:66      Can't connect to Redis. Retrying        {"error": "dial tcp 127.0.0.1:6380: connect: connection refused"}
2021-09-22T16:28:19.135+0200    INFO    config/redis.go:71      Re-connected to Redis   {"after": "16.001105518s", "attempts": 30}
```

**Comment by Al2Klimov** [comment]
```
2021-09-22T16:32:54.077+0200    INFO    icingadb/main.go:45     Connecting to database
2021-09-22T16:32:54.077+0200    WARN    driver/driver.go:39     Can't connect to database. Retrying     {"error": "dial tcp 127.0.0.1:6033: connect: connection refused"}
2021-09-22T16:33:05.357+0200    DEBUG   driver/driver.go:80     [unexpected EOF]
2021-09-22T16:33:05.357+0200    WARN    driver/driver.go:39     Can't connect to database. Retrying     {"error": "driver: bad connection"}
2021-09-22T16:33:16.202+0200    INFO    driver/driver.go:44     Re-connected to database        {"after": "22.125350414s", "attempts": 9}
```

File: pkg/retry/retry.go [hunk 1]
```diff
@@ -13,25 +13,43 @@ type RetryableFunc func(context.Context) error
...
+	// OnSuccess is called once the operation succeeds.
+	OnSuccess func(elapsed time.Duration, attempt uint64, lastErr error)
+}
+
 // WithBackoff retries the passed function if it fails and the error allows it to retry.
 // The specified backoff policy is used to determine how long to sleep between attempts.
-// Once the specified timeout (if >0) elapses, WithBackoff gives up.
 func WithBackoff(
-	ctx context.Context, retryableFunc RetryableFunc, retryable IsRetryable, b backoff.Backoff, timeout time.Duration,
+	ctx context.Context, retryableFunc RetryableFunc, retryable IsRetryable, b backoff.Backoff, settings Settings,
```
Comment (lippserd) [inline pkg/retry/retry.go:29]: Any reason to not use a pointer for `settings`?

**Review by lippserd — CHANGES_REQUESTED** [review]

File: pkg/retry/retry.go (hunk 1 above)
Comment (Al2Klimov) [reply pkg/retry/retry.go:29]: Go would allocate the `&Settings{}` on the stack, too.

**Review by lippserd — APPROVED** [review]

**Review by julianbrost — APPROVED** [review]

**Pull request merged**